File: benchmarks/tasks/wordfreq_arena/wordfreq_arena.js
// `wordfreq`, with an arena: the same counts, with nothing allocated per word.
//
// The corpus is read once into one Buffer, and a word is a pair of numbers into
// it -- where it starts and how long it is -- so no word is ever copied. The
// table is three typed arrays of a fixed size rather than a Map, and probing
// writes numbers into them. What is left in the number is the hashing, the
// probing and the byte comparisons, with the allocator and the string type
// taken out.
//
// Only the ten reported at the end become strings.

const fs = require("fs");

const CAP = 1 << 14; // 16384 slots for a vocabulary of 5000
const starts = new Int32Array(CAP);
const lens = new Int32Array(CAP);
const counts = new Float64Array(CAP);
const text = fs.readFileSync("work/corpus.txt");

// FNV-1a, in 32 bits: JavaScript has no 64-bit integer arithmetic that is
// quick, and a 32-bit hash probes just as well over 16384 slots.
function hashOf(at, n) {
  let h = 2166136261;
  for (let i = 0; i < n; i++) {
    h ^= text[at + i];
    h = Math.imul(h, 16777619);
  }
  return h >>> 0;
}

function same(a, b, n) {
  for (let i = 0; i < n; i++) if (text[a + i] !== text[b + i]) return false;
  return true;
}

function bump(at, n) {
  let i = hashOf(at, n) & (CAP - 1);
  for (;;) {
    if (counts[i] === 0) {
      starts[i] = at;
      lens[i] = n;
      counts[i] = 1;
      return;
    }
    if (lens[i] === n && same(starts[i], at, n)) {
      counts[i]++;
      return;
    }
    i = (i + 1) & (CAP - 1);
  }
}

// Is the word in slot a before the one in slot b, count first then bytes?
function before(a, b) {
  if (counts[a] !== counts[b]) return counts[a] > counts[b];
  const n = Math.min(lens[a], lens[b]);
  for (let i = 0; i < n; i++) {
    const x = text[starts[a] + i], y = text[starts[b] + i];
    if (x !== y) return x < y;
  }
  return lens[a] < lens[b];
}

for (let i = 0; i < text.length; ) {
  while (i < text.length && (text[i] === 32 || text[i] === 10)) i++;
  const start = i;
  while (i < text.length && text[i] !== 32 && text[i] !== 10) i++;
  if (i > start) bump(start, i - start);
}

// The ten commonest, kept in order as the table is walked.
const top = [];
for (let i = 0; i < CAP; i++) {
  if (counts[i] === 0) continue;
  let at = top.length;
  while (at > 0 && before(i, top[at - 1])) at--;
  if (at >= 10) continue;
  top.splice(at, 0, i);
  if (top.length > 10) top.length = 10;
}
console.log(
  top
    .map((i) => `${text.toString("latin1", starts[i], starts[i] + lens[i])}:${counts[i]}`)
    .join(" "),
);

File: benchmarks/tasks/wordfreq_arena/Wordfreq_arena.java
// `wordfreq`, with an arena: the same counts, with nothing allocated per word.
//
// The corpus is read once into one byte[], and a word is a pair of numbers into
// it -- where it starts and how long it is -- so no word is ever copied. The
// table is three flat arrays of a fixed size rather than a HashMap<String, ?>,
// and probing writes numbers into them. What is left in the number is the
// hashing, the probing and the byte comparisons, with the allocator and the
// string type taken out.
//
// Only the ten reported at the end become strings.

import java.nio.file.Files;
import java.nio.file.Path;

public class Wordfreq_arena {
    static final int CAP = 1 << 14; // 16384 slots for a vocabulary of 5000

    static final int[] starts = new int[CAP];
    static final int[] lens = new int[CAP];
    static final long[] counts = new long[CAP];
    static byte[] text;

    static long hashOf(int at, int n) {
        long h = 1469598103934665603L;
        for (int i = 0; i < n; i++) {
            h ^= (text[at + i] & 0xff);
            h *= 1099511628211L;
        }
        return h;
    }

    static boolean same(int a, int b, int n) {
        for (int i = 0; i < n; i++) if (text[a + i] != text[b + i]) return false;
        return true;
    }

    static void bump(int at, int n) {
        int i = (int) (hashOf(at, n) & (CAP - 1));
        for (;;) {
            if (counts[i] == 0) {
                starts[i] = at; lens[i] = n; counts[i] = 1;
                return;
            }
            if (lens[i] == n && same(starts[i], at, n)) { counts[i]++; return; }
            i = (i + 1) & (CAP - 1);
        }
    }

    // Is the word in slot a before the one in slot b, count first then bytes?
    static boolean before(int a, int b) {
        if (counts[a] != counts[b]) return counts[a] > counts[b];
        int n = Math.min(lens[a], lens[b]);
        for (int i = 0; i < n; i++) {
            int x = text[starts[a] + i] & 0xff, y = text[starts[b] + i] & 0xff;
            if (x != y) return x < y;
        }
        return lens[a] < lens[b];
    }

    public static void main(String[] args) throws Exception {
        text = Files.readAllBytes(Path.of("work/corpus.txt"));

        for (int i = 0; i < text.length;) {
            while (i < text.length && (text[i] == ' ' || text[i] == '\n')) i++;
            int start = i;
            while (i < text.length && text[i] != ' ' && text[i] != '\n') i++;
            if (i > start) bump(start, i - start);
        }

        // The ten commonest, kept in order as the table is walked.
        int[] top = new int[10];
        int n = 0;
        for (int i = 0; i < CAP; i++) {
            if (counts[i] == 0) continue;
            int at = n;
            while (at > 0 && before(i, top[at - 1])) at--;
            if (at >= 10) continue;
            for (int j = Math.min(n, 9); j > at; j--) top[j] = top[j - 1];
            top[at] = i;
            if (n < 10) n++;
        }
        StringBuilder out = new StringBuilder();
        for (int i = 0; i < n; i++) {
            if (i > 0) out.append(' ');
            out.append(new String(text, starts[top[i]], lens[top[i]]))
               .append(':').append(counts[top[i]]);
        }
        System.out.println(out);
    }
}
